### PR TITLE
cli: new flag to skip phases of upgrade

### PR DESCRIPTION
**Pull request by elchead**
### Context

Currently it is not possible to select / deselect specific upgrades (terraform, helm, k8s, node image).

### Proposed change(s)

- add a new `--skip-phases` flag that accepts a list of phases to skip during the upgrade

### Related issue
- [AB#3407](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3407)

### Checklist

- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: cli/internal/cmd/upgradeapply.go [hunk 1]
```diff
@@ -551,6 +593,20 @@ type upgradeApplyFlags struct {
...
+	SkipPhases        SkipPhases
+}
+
+// SkipPhases is a list of phases that can be skipped during the upgrade process.
+type SkipPhases []SkipPhase
+
+// Contains returns true if the list of phases contains the given phase.
+func (s SkipPhases) Contains(phase SkipPhase) bool {
+	for _, p := range s {
+		if p == phase {
```
Comment (daniel-weisse) [inline cli/internal/cmd/upgradeapply.go:605]:
```suggestion
		if strings.EqualFold(p, phase) {
```
Use [`strings.EqualFold`](https://pkg.go.dev/strings#EqualFold) to have a reliable way for comparing case-insensitive strings.

File: cli/internal/cmd/upgradeapply.go [hunk 2]
```diff
@@ -551,6 +594,20 @@ type upgradeApplyFlags struct {
...
 	conformance       bool
 	helmWaitMode      helm.WaitMode
+	skipPhases        skipPhases
+}
+
+// skipPhases is a list of phases that can be skipped during the upgrade process.
+type skipPhases []skipPhase
+
+// Contains returns true if the list of phases contains the given phase.
+func (s skipPhases) Contains(phase skipPhase) bool {
```
Comment (daniel-weisse) [inline cli/internal/cmd/upgradeapply.go:604]:
```suggestion
// contains returns true if the list of phases contains the given phase.
func (s skipPhases) contains(phase skipPhase) bool {
```

**Review by daniel-weisse — APPROVED** [review]
Can you add a ticket to document this feature?

**Review by malt3 — APPROVED** [review]

File: cli/internal/cmd/upgradeapply.go [hunk 3]
```diff
@@ -37,6 +38,20 @@ import (
...
 
+const (
+	// skipInfrastructurePhase skips the terraform apply of the upgrade process.
+	skipInfrastructurePhase skipPhase = "infrastructure"
+	// skipHelmPhase skips the helm upgrade of the upgrade process.
+	skipHelmPhase skipPhase = "helm"
+	// skipNodePhase skips the node upgrade of the upgrade process.
+	skipNodePhase skipPhase = "node"
+	// skipK8sPhase skips the k8s upgrade of the upgrade process.
+	skipK8sPhase skipPhase = "k8s"
```
Comment (thomasten) [inline cli/internal/cmd/upgradeapply.go:49]: this is not implemented

File: cli/internal/cmd/upgradeapply.go [hunk 4]
```diff
@@ -53,6 +68,7 @@ func newUpgradeApplyCmd() *cobra.Command {
 		"Might be useful for slow connections or big clusters.")
 	cmd.Flags().Bool("conformance", false, "enable conformance mode")
 	cmd.Flags().Bool("skip-helm-wait", false, "install helm charts without waiting for deployments to be ready")
+	cmd.Flags().StringSlice("skip-phases", []string{}, "skip one or multiple phases of the upgrade process {infrastructure|helm|node|k8s}\n")
```
Comment (thomasten) [inline cli/internal/cmd/upgradeapply.go:71]:
```suggestion
	cmd.Flags().StringSlice("skip-phases", nil, "comma-separated list of upgrade phases to skip\n"+
		"one or multiple of { infrastructure | helm | node | k8s }")
```
optional: I'd make it a bit more clear how to pass these values

File: cli/internal/cmd/upgradeapply.go [hunk 5]
```diff
@@ -509,6 +536,21 @@ func parseUpgradeApplyFlags(cmd *cobra.Command) (upgradeApplyFlags, error) {
 	if skipHelmWait {
 		helmWaitMode = helm.WaitModeNone
 	}
+
+	rawSkipPhases, err := cmd.Flags().GetStringSlice("skip-phases")
+	if err != nil {
+		return upgradeApplyFlags{}, fmt.Errorf("parsing skip-phases flag: %w", err)
+	}
+	skipPhases := []skipPhase{}
```
Comment (thomasten) [inline cli/internal/cmd/upgradeapply.go:544]:
```suggestion
	var skipPhases []skipPhase
```
nit: [nil is a valid slice](https://github.com/edgelesssys/wiki/blob/master/company/gostyle.md#nil-is-a-valid-slice)

**Review by thomasten — CHANGES_REQUESTED** [review]

**Review by thomasten — APPROVED** [review]
Code LGTM. (I haven't tried it out.)

**Review by daniel-weisse — APPROVED** [review]

**Pull request merged**